dhash/module.py

```diff
@@ -0,0 +1,508 @@
+
+                # do the comparison
+                db_image_hash = int(image.hash, 16)
+                distance = dhash.get_num_bits_different(db_image_hash, image_hash)
+                if distance < minimal_distance:
+                    duplicate = image
+                    minimal_distance = distance
+
+            if minimal_distance < LIMIT_SOFT:
+                duplicates[duplicate] = minimal_distance
+
+        for image_hash, distance in duplicates.items():
+            await self._report_duplicate(message, image_hash, distance)
+
+    async def _report_duplicate(
+        self, message: discord.Message, original: ImageHash, distance: int
+    ):
+        """Send report.
+        message: The new message containing attachment repost.
+        original: The original attachment.
+        distance: Hamming distance between the original and repost.
+        """
+        tc = TranslationContext(message.guild.id, message.author.id)
+
+        if distance <= LIMIT_FULL:
+            level = _(tc, "**♻ This is repost!**")
+        elif distance <= LIMIT_HARD:
+            level = _(tc, "**♻ This is probably repost!**")
+        else:
+            level = _(tc, "🤷🏻 This could be repost.")
+
+        await message.add_reaction("♻")
+
+        similarity = "{:.1f} %".format((1 - distance / 128) * 100)
+        timestamp = utils.Time.id_to_datetime(original.attachment_id).strftime(
+            "%Y-%m-%d %H:%M:%S"
+        )
+
+        try:
+            original_channel = message.guild.get_channel(original.channel_id)
+            original_message = await original_channel.fetch_message(original.message_id)
+            author = discord.utils.escape_markdown(original_message.author.display_name)
+            link = f"[**{author}**, {timestamp}]({original_message.jump_url})"
+        except discord.errors.NotFound:
+            link = "404 😿"
+
+        description = _(tc, "{name}, matching **{similarity}**!").format(
+            name=discord.utils.escape_markdown(message.author.display_name),
+            similarity=similarity,
+        )
+
+        embed = utils.Discord.create_embed(title=level, description=description)
+
+        embed.add_field(
+            name=_(tc, "Original"),
+            value=link,
+            inline=False,
+        )
+
+        embed.add_field(
+            name=_(tc, "Hint"),
+            value=_(
+                tc,
+                "_If image is repost, give it ♻️ reaction. "
+                "If it's not, click here on ❎ and when we reach {limit} reactions, "
+                "this message will be deleted._",
+            ).format(
+                limit=HashChannel.get(
+                    message.guild.id, message.channel.id
+                ).reaction_limit,
+            ),
+            inline=False,
+        )
+        embed.set_footer(text=f"{message.author.id} | {message.id}")
+
+        report = await message.reply(embed=embed)
+
+        self.embed_cache[message.id] = report.id
```

Comment:
I wonder -- can we save full `report` as a value, instead of its ID only? We could save ourselves one message `.fetch()` in the logic.

Reply:
Good idea, gonna do that.